### PR TITLE
feat: $ref cannot have siblings

### PR DESCRIPTION
**Pull request by P0lip**
Closes https://github.com/stoplightio/spectral/issues/394

**Checklist**

- [x] Tests added / updated
- [x] Docs added / updated

**Does this PR introduce a breaking change?**

- [ ] Yes
- [x] No

It's not a rule, as rules have no access to original (unresolved) data.
I don't know if it's a big deal or not, but if we do need a rule, I can make it a rule.
The code is placed in a wrong place and an extra test wouldn't hurt, so it's still a WIP.

### CODE REVIEW

**Comment by P0lip** [comment]
@philsturgeon 
> It's not a rule, as rules have no access to original (unresolved) data.
> I don't know if it's a big deal or not, but if we do need a rule, I can make it a rule.

How do you feel about not making it a rule and failing all the time same as invalid-ref does?

**Comment by P0lip** [comment]
@philsturgeon got it, will make it a rule then.

**Comment by P0lip** [comment]
@philsturgeon 
I'm planning to add an extra property to a rule named `unresolved` or something, which would make each rule operate on initial data

**Comment by P0lip** [comment]
Hold on with the review, forgot to add one test.

File: src/types/rule.ts [hunk 1]
```diff
@@ -26,6 +26,9 @@ export interface IRule<T = string, O = any> {
   // Filter the target down to a subset[] with a JSON path
   given: string;
 
+  // If true, rule will operate on original (unresolved) data
+  unresolved?: boolean;
```
Comment (philsturgeon) [inline src/types/rule.ts]: Would it be better to swap? `unresolved: false` sounds a bit awkward and double-negative. Would something like `isResolved: true` be easier for others to read and reason about?

File: rollup.config.js [hunk 2]
```diff
@@ -7,22 +7,21 @@ import { terser } from 'rollup-plugin-terser';
...
+  'refSiblings'
 ].map(fn => ({
   input: path.resolve(BASE_PATH, 'dist/rulesets/oas/functions', `${fn}.js`),
   plugins: [
     typescript({
       tsconfig: path.join(BASE_PATH, './tsconfig.rollup.json'),
       include: ['dist/**/*.{ts,tsx}'],
     }),
-    resolve({
-      only: ['json-schema-merge-allof', /lodash\/?.*/],
```
Comment (philsturgeon) [inline rollup.config.js]: I'm not sure what `json-schema-merge-allof` but is it a dependency we can remove now?

Comment (P0lip) [reply rollup.config.js]: @philsturgeon actually it's a leftover :laughing: 
I copied that rollup config from json-schema-viewer repo and forgot to remove it.
We don't have that package included in our package.json, so we should be okay.

File: src/rulesets/oas/index.json [hunk 3]
```diff
@@ -428,6 +429,18 @@
           "schemaPath": "$"
         }
       }
+    },
+    "$ref-siblings": {
+      "description": "$ref cannot be placed among any other property",
```
Comment (philsturgeon) [inline src/rulesets/oas/index.json]:
```suggestion
      "description": "$ref cannot be placed next to any other properties",
```

File: src/rulesets/oas/index.json [hunk 4]
```diff
@@ -428,6 +429,18 @@
           "schemaPath": "$"
         }
       }
+    },
+    "$ref-siblings": {
```
Comment (philsturgeon) [inline src/rulesets/oas/index.json]:
```suggestion
    "no-$ref-siblings": {
```

File: src/types/rule.ts [hunk 5]
```diff
@@ -26,6 +26,10 @@ export interface IRule<T = string, O = any> {
   // Filter the target down to a subset[] with a JSON path
   given: string;
 
+  // If false, rule will operate on original (unresolved) data
```
Comment (philsturgeon) [inline src/types/rule.ts]: It seems like this will update our docs a bit? Could you add mention of it to the custom rules section?

Comment (P0lip) [reply src/types/rule.ts]: @philsturgeon added a short note, lmk what you think.

File: docs/getting-started/rulesets.md [hunk 6]
```diff
@@ -40,6 +40,24 @@ While running it with this object, it will succeed:
...
+By default, Spectral processes each rule on resolved document with all $refs resolved.
+If you would like to have an original input supplied to your rule, you can place `isResolved` property as follows:
+
+```yaml
+rules:
+  my-rule-name:
+    description: Tags must have a description.
+    given: $.tags[*]
+    severity: error
+    isResolved: false # note - if not specified or true, a resolved document will be given
```
Comment (philsturgeon) [inline docs/getting-started/rulesets.md]: I'm sorry this is a switch and not a result flag. I missed it before but now it makes sense. Can we make it `resolved: true/false`?  😅

**Pull request merged**